mm2src/mm2_bitcoin/keys/src/segwitaddress.rs

```diff
@@ -221,4 +226,53 @@ mod tests {
         );
         assert_eq!(addr.address_type(), Some(AddressType::P2wsh));
     }
+
+    #[test]
+    // https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#test-vectors
+    fn test_invalid_segwit_addresses() {
+        // Invalid checksum
+        let invalid_address = "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t5";
+        let err = SegwitAddress::from_str(invalid_address).unwrap_err();
+        assert_eq!(err, Error::Bech32(bech32::Error::InvalidChecksum));
+
+        // Invalid witness version
+        let invalid_address = "BC13W508D6QEJXTDG4Y5R3ZARVARY0C5XW7KN40WF2";
+        let err = SegwitAddress::from_str(invalid_address).unwrap_err();
+        assert_eq!(err, Error::InvalidWitnessVersion(17));
+
+        // Invalid program length
+        let invalid_address = "bc1rw5uspcuh";
+        let err = SegwitAddress::from_str(invalid_address).unwrap_err();
+        assert_eq!(err, Error::InvalidWitnessProgramLength(1));
+
+        // Invalid program length
+        let invalid_address = "bc10w508d6qejxtdg4y5r3zarvary0c5xw7kw508d6qejxtdg4y5r3zarvary0c5xw7kw5rljs90";
+        let err = SegwitAddress::from_str(invalid_address).unwrap_err();
+        assert_eq!(err, Error::InvalidWitnessProgramLength(41));
+
+        // Invalid program length for witness version 0 (per BIP141)
+        let invalid_address = "BC1QR508D6QEJXTDG4Y5R3ZARVARYV98GJ9P";
+        let err = SegwitAddress::from_str(invalid_address).unwrap_err();
+        assert_eq!(err, Error::InvalidSegwitV0ProgramLength(16));
+
+        // Mixed case invalid
+        let invalid_address = "tb1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3q0sL5k7";
+        let err = SegwitAddress::from_str(invalid_address).unwrap_err();
+        assert_eq!(err, Error::Bech32(bech32::Error::MixedCase));
+
+        // zero padding of more than 4 bits
+        let invalid_address = "bc1zw508d6qejxtdg4y5r3zarvaryvqyzf3du";
+        let err = SegwitAddress::from_str(invalid_address).unwrap_err();
+        assert_eq!(err, Error::Bech32(bech32::Error::InvalidPadding));
+
+        // Non-zero padding in 8-to-5 conversion
+        let invalid_address = "tb1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3pjxtptv";
+        let err = SegwitAddress::from_str(invalid_address).unwrap_err();
+        assert_eq!(err, Error::Bech32(bech32::Error::InvalidPadding));
+
+        // Empty data section
+        let invalid_address = "bc1gmk9yu";
+        let err = SegwitAddress::from_str(invalid_address).unwrap_err();
+        assert_eq!(err, Error::EmptyBech32Payload);
+    }
 }
```

Comment:
Could you please also add test covering valid cases from BIP-0173 vectors? Like
```
    #[test]
    fn test_valid_segwit() {
        use hex::ToHex;
        let addr = "BC1QW508D6QEJXTDG4Y5R3ZARVARY0C5XW7KV8F3T4";
        let segwit_addr = SegwitAddress::from_str(addr).unwrap();
        assert_eq!(
            "751e76e8199196d454941c45d1b3a323f1433bd6",
            segwit_addr.program.to_hex::<String>()
        );
    }
```

Reply:
Not all valid cases in BIP-0173 vectors are valid segwit (segwit_v0) addresses, they are all valid bech32 address format but some use versions other than 0. BIP-0350 used version 1 for taproot and changed the way the scriptPubKey is calculated that's why they added a new checksum and new format called bech32m.

Reply:
Versions other than 0 shouldn't be used with bech32 format in bitcoin

Reply:
I have not noticed it. Surely, covering non-applicable cases was not required, thanks!